lightning/src/ln/channel.rs

```diff
@@ -4020,6 +4013,15 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 	}
 }
 
+fn is_unsupported_shutdown_script(their_features: &InitFeatures, scriptpubkey: &Script, is_outbound: bool) -> bool {
+	// BOLT 2 says we must only send a scriptpubkey of certain standard forms,
+	// which for a a BIP-141-compliant witness program is at max 42 bytes in length.
+	// So don't let the remote peer feed us some super fee-heavy script.
+	let is_script_too_long = is_outbound && scriptpubkey.len() > 42;
+	let not_valid_witness = !their_features.supports_shutdown_anysegwit() && scriptpubkey.is_witness_program();
```

Comment:
Shouldn't this be `!their_features.supports_shutdown_anysegwit() || !scriptpubkey.is_witness_program()`? That would mean the script is invalid if the script is too long, or ((its not a witness or they didn't say they wanted any witness) and its not an otherwise-known type).

Reply:
Indeed this doesn't seem right. `their_features.supports_shutdown_anysegwit() && !scriptpubkey.is_witness_program()` sounds better in my head...

Reply:
Oops, this the witness check is also incorrect - witness version 0 where the push size is not 20 or 32 is non-standard so would result in us signing a non-standard transaction. The BOLT PR captures this, noting that anysegwit enables " `OP_1` through `OP_16` inclusive, followed by a single push of 2 to 40 bytes" but *not* `OP_0`.

Reply:
I'll add a specific test to make sure `OP_0` starting script fails and will adjust other test to make sure `OP_1` or any following are allowed.